File: Leetcode/Trees and Graphs/decode_string.py
class Solution:
    def decode_string(self, s):
        stack = []
        curr_num = 0
        curr_string = ''

        for char in s:
            if char == '[':
                stack.append(curr_string)
                stack.append(curr_num)
                curr_string = ''
                curr_num = 0
            elif char == ']':
                num = stack.pop()
                prev_string = stack.pop()
                curr_string = prev_string + (num*curr_string)
            elif char.isdigit():
                curr_num = curr_num*10 + int(char)
            else:
                curr_string += char

        return curr_string

# Time Complexity: O(n)
# Space Complexity: O(n)

File: Cheetcode Special Problems/decodeString.js
const decodeString = (s) => {
  const stack = [];
  let currNum = 0;
  let currString = "";

  for (let i = 0; i < s.length; i++) {
    const ele = s[i];
    if (ele === "[") {
      stack.push(currString);
      stack.push(currNum);
      currString = ""
      currNum = 0
    } else if (ele === "]") {
      const num = stack.pop();
      const prevString = stack.pop();
      currString = prevString + (currString.repeat(num))
    } else if (!Number.isNaN(Number(ele))) {
      currNum = (currNum * 10) + Number(ele)
    } else {
      currString += ele
    }
  }

  return currString;
}

// Time Complexity: O(N * maxK^countK);
// Space Complexity: O(sum(maxK^K * N));